javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,176 @@
+import arcs.android.crdt.ParcelableCrdtType
+import arcs.android.storage.service.testing.FakeMessageCallback
+import arcs.android.storage.service.testing.FakeStorageChannelCallback
+import arcs.android.storage.toParcelToByteArray
+import arcs.core.data.CountType
+import arcs.core.storage.FixedDriverFactory
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StorageKeyParser
+import arcs.core.storage.StoreOptions
+import arcs.core.storage.UntypedProxyMessage
+import arcs.core.storage.keys.RamDiskStorageKey
+import arcs.core.storage.testutil.MockDriverProvider
+import arcs.core.storage.testutil.testWriteBackProvider
+import arcs.core.type.Type
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class StorageServiceNgImplTest {
+
+  private val driverFactory = FixedDriverFactory(MockDriverProvider())
+  private val onProxyMessageCallback: suspend (StorageKey, UntypedProxyMessage) -> Unit =
+    { _: StorageKey, _: UntypedProxyMessage -> }
+  private val messageCallback = FakeMessageCallback()
+  private val channelCallback = FakeStorageChannelCallback()
+
+  private val scope = TestCoroutineScope()
+  private lateinit var storageService: StorageServiceNgImpl
+
+  @Before
+  fun setUp() {
+    StorageKeyParser.addParser(RamDiskStorageKey)
+    storageService = StorageServiceNgImpl(
+      scope,
+      driverFactory,
+      ::testWriteBackProvider,
+      null,
+      onProxyMessageCallback
+    )
+  }
+
+  @After
+  fun tearDown() {
+    scope.cleanupTestCoroutines()
+  }
+
+  @Test
+  fun openStorageChannel_invokesOnCreateCallback_andCreatesChannel() = scope.runBlockingTest {
+    val encodedStoreOptions = createEncodedStoreOptions(
+      RamDiskStorageKey("myCount"),
+      CountType(),
+      ParcelableCrdtType.Count
+    )
+
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+
+    // `waitForOnCreate` ensures the `onCreate` callback was invoked.
+    val channel = channelCallback.waitForOnCreate()
+
+    assertThat(channel).isInstanceOf(StorageChannelImpl::class.java)
+  }
+
+  @Test
+  fun openingTwoChannels_forSameStore_resultsInTwoDistinctChannels() = scope.runBlockingTest {
+    val encodedStoreOptions = createEncodedStoreOptions(
+      RamDiskStorageKey("myCount"),
+      CountType(),
+      ParcelableCrdtType.Count
+    )
+
+    val channelCallback2 = FakeStorageChannelCallback()
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback2, messageCallback)
+
+    val channel1 = channelCallback.waitForOnCreate()
+    val channel2 = channelCallback2.waitForOnCreate()
+    assertThat(channel2).isNotSameInstanceAs(channel1)
```

Comment:
It might also be worth checking that `channel1.store` is the same as `channel2.store`. As in, check that they are different channels pointing to an identical store.

Reply:
This is tested in `openingTwoChannels_forSameStore_onlyCreatesOneStore`. I considered merging the two tests but I decided against it in favour keeping tests simple and only testing one thing at a time. Also the test name would be something like: `openingTwoChannels_forSameStore_createsTwoChannel_but_onlyCreatesOneStore` which is a bit much 😄